gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/metastore/HiveMetaStoreUtils.java

```diff
@@ -261,27 +259,69 @@ public static SerDeInfo getSerDeInfo(HiveRegistrationUnit unit) {
     return si;
   }
 
+  public static boolean containsNonOptionalUnionTypeColumn(Table t) {
+    return containsNonOptionalUnionTypeColumn(getHiveTable(t));
+  }
+
   /**
-   * Util for detecting if {@param hiveTable} has a non-optional union (aka complex unions) column types. A non optional
-   * union is defined as a uniontype with multiple possible types and none of them are null
+   * Util for detecting if a hive table has a non-optional union (aka complex unions) column types. A non optional
+   * union is defined as a uniontype with n >= 2 non-null subtypes
    *
-   * @param hiveTable
-   * @return if hive table contains complex uniontype columns
+   * @param hiveTable Hive table
+   * @return if hive table contains non-optional uniontype columns
    */
   public static boolean containsNonOptionalUnionTypeColumn(HiveTable hiveTable) {
-    if (!isAvroFormat(hiveTable)) {
-      // All values in ORC are optional / nullable
-      return false;
+    if (hiveTable.getProps().contains("avro.schema.literal")) {
+      Schema.Parser parser = new Schema.Parser();
+      Schema schema = parser.parse(hiveTable.getProps().getProp("avro.schema.literal"));
+      return isNonOptionalUnion(schema);
     }
 
-    return hiveTable.getColumns().stream()
-        .map(HiveRegistrationUnit.Column::getType)
-        .map(Object::toString)
-        .anyMatch(columnType -> columnType.contains("uniontype") && !columnType.contains("void"));
+    if (isNonAvroFormat(hiveTable)) {
```

Comment:
This is a fallback case if schema literal is not set. Where we can use the ORC type parser to determine if the column is a non-optional union.

This does not work if the underlying table is not ORC based